File: week_8_and_9/4_Ruby/fizz_buzz/my_solution.rb
# U3.W8-9: 


# I worked on this challenge by myself.

# 2. Pseudocode

# define super_fizzbuzz (array)
# map array
# if element % 15 == 0 
#    fizzbuzz
# if element % 3 == 0
#    fizz
# if element % 5 == 0
#    buzz
# else 
#    element
# ENDIF

# 3. Initial Solution

def super_fizzbuzz(array)
  # Your code goes here!
  array.map do |x|
      if x % 15 == 0   
          "FizzBuzz"
      elsif
         x % 3 == 0
          "Fizz"
      elsif 
         x % 5 == 0
          "Buzz"
      else
         x  
      end
  end
end

# 4. Refactored Solution


# I'm not sure how to refactor my solution - if I did case, I would still need to write everything out, so I don't think that would be very efficient 
		



# 1. DRIVER TESTS/ASSERT STATEMENTS GO BELOW THIS LINE

def random_input_array(base, size)
  (1..size).map { |i| base**(1+rand(15)) }
end

p super_fizzbuzz(random_input_array(3,100)) === ["Fizz"]*100   
p super_fizzbuzz(random_input_array(5,100)) === ["Buzz"]*100
p super_fizzbuzz(random_input_array(15, 100)) === ["FizzBuzz"]*100
p super_fizzbuzz([1,2,3]) === [1,2,'Fizz']
p super_fizzbuzz([15, 5, 3, 1]) === ['FizzBuzz', 'Buzz', 'Fizz', 1]



# 5. Reflection 
# I have to be honest: I went back to my solution on Socrates for this one. I refactored it (instead of pushing the strings to an array, I just had the program map it to a temporary array). I enjoy fizzbuzz a lot - it's a logic problem. Not so much about syntax, just pure step-by-step logic. I like that stuff, and I remember when I figured this one out. I felt incredibly fantastic with it, and now as I look at it, it makes me happy that I can see what's actually happening and refactor it easily.

File: week_8_and_9/3_JavaScript/translate_to_js/my_solution.js
// U3.W8-9: Challenge you're converting : FIZZ BUZZ


// I worked on this challenge by myself

// 2. Pseudocode

// fizz_buzz (array)
// loop through the passed array
// if i is div. by 15
// FIZZBUZZ
// if i is div. by 3
// FIZZ
// if i is div by 5 
// BUZZ
// else pass i to array

// 3. Initial Solution
function superFizzBuzz(array){
	fizzArray = [];
	for (var i = 0; i < array.length; i++) {
		if (array[i] % 15 === 0) {
			fizzArray.push("FizzBuzz");
		}
		else if (array[i] % 3 === 0) {
			fizzArray.push("Fizz");
		}
		else if (array[i] % 5 === 0) {
			fizzArray.push("Buzz");
		}
		else fizzArray.push(array[i]);
	}
	return fizzArray;
}

// 4. Refactored Solution

// I'm not sure how to refactor this solution, as every part seems very necessary




// 1. DRIVER TESTS/ASSERT STATEMENTS GO BELOW THIS LINE
console.log(superFizzBuzz([3,15]));
console.log(superFizzBuzz([3,40, 35, 40, 10, 10]));
console.log(superFizzBuzz([10, 10, 10, 10, 10, 10, 10]));
console.log(superFizzBuzz([3,15]));
console.log(superFizzBuzz([3,15]));

// 5. Reflection 

// I enjoy the translation problems. What is frustrating me the most is driver tests in JS. I must have completely forgotten how to run them properly - I think I have just gotten so comfortable with Ruby that I can't think in JS correctly. So many brackets. SO MANY BRACKETS. Otherwise, this was fairly straight forward. A for loop, and indexing. Pretty typical JS work, and it was a simple one to hammer out!
